odl/trafos/fourier.py

```diff
@@ -0,0 +1,1024 @@
+        if not isinstance(dom.grid, RegularGrid):
+            raise TypeError('irregular grids not supported.')
+
+        if ran is None:
+            # Calculate range - a complex DiscreteLp with conjugate exponent
+            if dom.exponent == 1.0:
+                conj_exp = float('inf')
+            elif dom.exponent == float('inf'):
+                conj_exp = 1.0
+            else:
+                conj_exp = dom.exponent / (dom.exponent - 1.0)
+
+            # Standard grid with stride 1 and minimum (0, ..., 0)
+            ran_grid = RegularGrid([0] * dom.ndim, np.array(dom.shape) - 1,
+                                   dom.shape, as_midp=False)
+
+            ran_fspace = FunctionSpace(ran_grid.convex_hull(),
+                                       ComplexNumbers())
+
+            if is_real_dtype(dom.dtype):
+                ran_dtype = _TYPE_MAP_R2C[dom.dtype]
+            else:
+                ran_dtype = dom.dtype
+
+            ran_dspace_type = dspace_type(ran_fspace, impl='numpy',
+                                          dtype=ran_dtype)
+            ran_dspace = ran_dspace_type(ran_grid.size, dtype=ran_dtype,
+                                         exponent=conj_exp)
+            ran = DiscreteLp(ran_fspace, ran_grid, ran_dspace,
+                             exponent=conj_exp)
+
+        super().__init__(dom, ran, linear=True)
+        self._fftw_plan = None
+
+    def _call(self, x, out=None, **kwargs):
+        """Implementation of ``self(x[, out])``.
+
+        Parameters
+        ----------
+        x : domain element
+            Input vector to be transformed
+        out : range element, optional
+            Output vector storing the result
+        direction : {'FFTW_FORWARD', 'FFTW_BACKWARD'}, optional
+            Direction of the transform
+        normalise_idft : `bool`, optional
+            If `True`, the IDFT (``'FFTW_BACKWARD'``) is normalized by
+            ``1 / N``, where ``N`` is the total number of points in
+            ``x[axes]``. This ensures that the IDFT is the true inverse
+            of the forward DFT.
+            Default: `True`
+        flags : sequence of `str`, optional
+            Flags for the transform. See the `pyfftw API documentation`_
+            for futher information.
+            Default: ``('FFTW_MEASURE',)``
+        threads : positive `int`, optional
+            Number of threads to use. Default: 1
+        planning_timelimit : `float` or `None`, optional
+            Rough upper limit in seconds for the planning step of the
+            transform. `None` means no limit. See the
+            `pyfftw API documentation`_ for futher information.
+
+        Returns
+        -------
+        out : `DiscreteLpVector`
+            Result of the transform. If ``out`` was given, the returned
+            object is a reference to it.
+        """
+        if out is None:
+            out = self.range.element()
+
+        self._fftw_plan = pyfftw_call(
```

Comment:
The thing is that if `pyfftw_call` does FFT planning internally, the arrays are zeroed out. So once the plan is there and can be reused we're on the safe side. Otherwise, just checking if the data was once copied doesn't help much. Or am I getting something wrong?